### DISCUSSION TITLE
enum issues during development

### DISCUSSION THREAD

**Discussion by errael** [vim/vim#14278] [General]
@yegappan I'm starting this discussion since it seems better that muddying up the PR (and I could be wrong)

**Comment by errael** [comment]
## weirdness using this in `new()`
March 23

Not sure what the issue is. 
- Can't reference this in constructor. In which case there should be an error.
- Incomplete construction before `new()` is invoked

```
vim9script

enum Planet
    earth, venus
    def new()
        echo printf("new Planet: %s", this)
    enddef
endenum
```
outputs
```
new Planet: enum Planet.earth {ordinal: 0}
new Planet: enum Planet.earth {ordinal: 0}
```

A different example, 
```
vim9script

var count: number = 7

enum Planet
    earth, venus
    var x: number
    def new()
        this.x = count
        count += 1
        echo printf("new Planet: %s", this)
    enddef
endenum
```
```
new Planet: enum Planet.earth {ordinal: 0, x: 7}
new Planet: enum Planet.earth {ordinal: 0, x: 8}
```

**Reply by yegappan** [reply]
Thanks for trying this out and giving the feedback.  You can reference `this` in the enum constructor.  The `ordinal` object variables are initialized after all the enum objects are created.  So in the enum constructor, the ordinal value is not yet initialized and set to the default value (0).  When the enum object is converted to a string for `echo`, the ordinal value is used to get the object.  In the constructor, the ordinal value is still zero, so echo prints the first enum object.  But when you modify any other member variable in the constructor, the correct object will be used.

**Reply by errael** [reply]
The first example now works as expected.

The second example produces

```
Error detected while processing :source buffer=1:
line    7:
E1426: Object variable "x: number" is not final or const in enum "Planet"
```

**Reply by yegappan** [reply]
Yes.  As enums are constants, the object variables of an enum should be a constant or final and can be initialized only in the constructor.

**Reply by errael** [reply]
>  As enums are constants, the object variables of an enum
>  should be a constant or final

I don't understand why. Considering an enum as a class
```
class Planet
    public static final earth: Planet # = Planet.new()
    public static const values: list<Planet> # = [earth]

    public const oridinal: number 
endclass
```

It's clear that the above access restrictions are required.

But why must other/additional developer defined object variable be `const` or `final`?

This handy feature is available in other languages: `Java`, `Kotlin`, `Scala`

**Reply by yegappan** [reply]
I modeled this after the Dart language specification (https://dart.dev/language/enums) where only final instance variables are supported. 
The Dart specification made sense to me as enum values are constant objects and should not have modifiable state.  But as you mentioned, Java, Kotlin and Scala support non-final instance variables in an enum.  As the other languages in this list are more popular, users may expect that behavior.  I will update the PR to remove this restriction.

**Reply by errael** [reply]
> enum values are constant objects and should not have modifiable state

When considering workarounds if an enum couldn't have variables, one solution was to setup a dict keyed by the enum, with the value being a "companion" class holding the variables associated with the enum.

Is there an enum method/variable for obtaining the enum name as a string? Java has a method `name()` to return the declared enum  value. That's what you'd want to use for a dict key.

**Reply by yegappan** [reply]
I have updated the PR to allow non-final and non-const object variables in an enum.  I will update the PR to include the support for the `name` object variable.

**Comment by errael** [comment]
## Lower case enums are allowed
March 23

**Reply by yegappan** [reply]
I didn't see any strong reason to not support enum values starting with a lower case letter.

**Reply by errael** [reply]
Oops, I left out references.

I'm OK with lower case. I left out that in vim9class.txt:931 it says
```
An enum name must start with an uppercase letter.	*E1415*
```
and all the examples show upper case.

**Reply by chrisbra** [reply]
should we really allow lower case enums? I fear it could cause ambigiouty.

**Reply by yegappan** [reply]
The first character of an enum name (e.g. Planet) must be an upper case character.  The name of an enum
value (e.g. venus) in an enum can start with a lower case or upper case character.  This should not cause any
ambiguity as the enum value will be prefixed with the enum name (e.g. Planet.venus).

**Comment by errael** [comment]
## Variable name tracking issue
March 23 (may have nothing to do with enums)

```
vim9script

interface I
    var x: number
endinterface

var count: number = 1

enum Planet implements I
    earth, venus
    var x: number
endenum

# NO ERROR IF THE FOLLOWING LINE IS AT THE END
var x = Planet.venus

class C1 implements I
    var x: number
endclass

C1.new()
```
```
Error detected while compiling :source buffer=1[21]..function C1.new:
line    1:
E1168: Argument already declared in the script: x
```

**Reply by yegappan** [reply]
Yes.  This issue is not related to enums.  The duplicate variable check is done only when compiling classes, enums and functions.  At the script level, when declaring a new variable, the duplicate name check is not done.

**Reply by errael** [reply]
> At the script level, when declaring a new variable, the duplicate name check is not done

Should it be?

**Reply by errael** [reply]
I'm wondering if shadowing checking should only be between items in a given namespace. For example,
- between script variables and function variables for functions defined in the same script
- between class variables and method variables of the same class

Checking between "namespaces" seems like a problem; confusing, ambiguous

Uh, hmm, the following works. Should it? It also works when defined in a class.

```
vim9script

def Fscript()
    def Gscript()
        var x = "x in Gscript()"
        echo x
    enddef

    var x = "x in Fscript()"
    echo x
    Gscript()
enddef
Fscript()
```

**Reply by zzzyxwvut** [reply]
The outcome depends on whether a default constructor (an  
error) or a user-defined constructor (no error) is used.

```
class C1 implements I
    var x: number

    def new()
    enddef
endclass
```
It is described in [this issue](https://github.com/vim/vim/issues/13764).

**Comment by errael** [comment]
## error in `enum`'s `new() ` causes `Internal error`

This is using PR with `Add the 'name' variable to the enum values.`
```
vim9script

enum Planet
    earth
    def new()
        x = 123
    enddef
endenum
```
```
Error detected while compiling :source buffer=1[8]..function Planet.new:
line    1:
E1100: Command not supported in Vim9 script (missing :var?): x = 123
Error detected while processing :source buffer=1:
line    8:
E340: Internal error; if you can reproduce please report a bug
E685: Internal error: copy_tv(UNKNOWN)
```

**Reply by yegappan** [reply]
I have updated the PR to handle this case.

**Comment by errael** [comment]
## duplicate `enum` value name is not an error

Expect and error from the following
```
vim9script

enum Planet
    earth, earth
    def new()
        echo this.name
    enddef
endenum
```
outputs
```
earth
earth
```

**Reply by yegappan** [reply]
I have updated the PR to handle this case.

**Comment by errael** [comment]
## Enum instance variable can not be initialized from a static.

After https://github.com/yegappan/vim/commit/f54d0dd782228eb8b33b0efce052b8801abed1af

Is executing the `new()` methods deferred until after the statics are initialized?

```
vim9script

enum Planet
    earth
    var xxx = Planet.values
endenum
```

```
Error detected while processing :source buffer=1[6]..function Planet.new:
line    1:
E340: Internal error; if you can reproduce please report a bug
E685: Internal error: copy_tv(UNKNOWN)
```

**Reply by errael** [reply]
An additional, simpler?, example of "Enum instance variable can not be initialized from a static."

```
enum Planet
    earth
    static var xxx = 33
    var yyy = xxx;
endenum
```

Notice the following does not get an error.
```
class C
    static var xxx = 33
    var yyy = xxx;
endclass
```

**Reply by yegappan** [reply]
The class variables are initialized one at a time starting from the enum value class variables.  To initialize the enum value class variable, the constructor is invoked.  In the above examples, this constructor is trying to initialize an object variable using a class variable which is not yet initialized.  This leads to the internal error.  I will see how to disallow this or handle this gracefully.

**Reply by errael** [reply]
Opened #14337.

**Comment by errael** [comment]
The enum implementation has been checked in so problems should go into "issues".
